domain/runaway.go

```diff
@@ -0,0 +1,625 @@
+			return
+		}
+
+		rows, sqlErr := do.execRestrictedSQL(sql, nil)
+		if sqlErr != nil {
+			logutil.BgLogger().Error("delete system table failed", zap.String("table", tableName), zap.Error(err))
+			return
+		}
+		leftRows = make([][]types.Datum, len(rows))
+		for i, row := range rows {
+			leftRows[i] = row.GetDatumRow(tb.KeyColumnTypes)
+		}
+
+		for len(leftRows) > 0 {
+			var delBatch [][]types.Datum
+			if len(leftRows) < runawayRecordGCBatchSize {
+				delBatch = leftRows
+				leftRows = nil
+			} else {
+				delBatch = leftRows[0:runawayRecordGCBatchSize]
+				leftRows = leftRows[runawayRecordGCBatchSize:]
+			}
+			sql, err := sqlbuilder.BuildDeleteSQL(tb, delBatch, expiredTime)
+			if err != nil {
+				logutil.BgLogger().Error(
+					"build delete SQL failed when deleting system table",
+					zap.Error(err),
+					zap.String("table", tb.Schema.O+"."+tb.Name.O),
+				)
+				return
+			}
+
+			_, err = do.execRestrictedSQL(sql, nil)
+			if err != nil {
+				logutil.BgLogger().Error(
+					"delete SQL failed when deleting system table", zap.Error(err), zap.String("SQL", sql),
+				)
+			}
+		}
+	}
+}
+
+func (do *Domain) runawayWatchSyncLoop() {
+	defer util.Recover(metrics.LabelDomain, "runawayWatchSyncLoop", nil, false)
+	runawayWatchSyncTicker := time.NewTicker(runawayWatchSyncInterval)
+	updateFn := func() error {
+		records, err := do.runawaySyncer.getNewWatchRecords()
+		if err != nil {
+			logutil.BgLogger().Error("try to get new runaway watch", zap.Error(err))
+			return err
+		}
+		for _, r := range records {
+			do.runawayManager.AddWatch(r)
+		}
+		doneRecords, err := do.runawaySyncer.getNewWatchDoneRecords()
+		if err != nil {
+			logutil.BgLogger().Error("try to get done runaway watch", zap.Error(err))
+			return err
+		}
+		for _, r := range doneRecords {
+			do.runawayManager.RemoveWatch(r)
+		}
+		return nil
+	}
+	for {
+		select {
+		case <-do.exit:
+			return
+		case <-do.runawaySyncer.notifyChan:
+			err := updateFn()
+			do.runawaySyncer.doneChan <- err
+		case <-runawayWatchSyncTicker.C:
+			err := updateFn()
+			if err != nil {
+				logutil.BgLogger().Warn("get runaway watch record failed", zap.Error(err))
+			}
+		}
+	}
+}
+
+// AddRunawayWatch is used to add runaway watch item manually.
+func (do *Domain) AddRunawayWatch(record *resourcegroup.QuarantineRecord) (int64, error) {
+	if err := do.handleRunawayWatch(record); err != nil {
+		return 0, err
+	}
+	do.runawaySyncer.notifyChan <- struct{}{}
+	timer := time.NewTimer(10 * time.Second)
+	ticker := time.NewTicker(time.Second)
+	defer func() {
+		ticker.Stop()
+		timer.Stop()
+	}()
+	for {
+		select {
+		case <-timer.C:
+			return 0, errors.Errorf("the query watch is added successfully, but the TiDB server load timed out, you can query information_schame.runaway_watches later to confirm")
+		case <-ticker.C:
+			r := do.runawayManager.GetWatchByKey(record.GetRecordKey())
+			if r.ID > 0 {
+				return r.ID, nil
+			}
+		case err := <-do.runawaySyncer.doneChan:
```

Comment:
What if multiple add watches happen at the same time, how can you ensure you get the right message from this chan?

Reply:
Can't ensure. If keys of records are different and it gets wrong message , It can't get watch and will try wo get watch later. If keys are same, refer to https://github.com/pingcap/tidb/pull/45465/files#r1277100060

Reply:
As you only send a message to notifyChan only once, one routine receives another routine's return message and drops it, then the other routine is blocked, this is a more severe situation.